// progres-convenience/index.js
"use strict";

var progres = require('progres');
var Q = require('q');


// Monkey-patch ProgresClient with convenience methods.


progres.ProgresClient.prototype.queryGenerated = function (generated) {

	var THIS = this;

	// Wrap with Q().then to make exceptions work.
	return Q().then(function () {

		var SQLAndValues = generated.toQuery();

		return THIS.query(SQLAndValues.text, SQLAndValues.values);
	});
};


progres.ProgresClient.prototype.insert = function (tableDefinition, objectOrArray) {

	var THIS = this;

	return Q().then(function () {
	
		return THIS.queryGenerated(
			tableDefinition
				.insert(objectOrArray)
		);
	});
};


progres.ProgresClient.prototype.select = function (tableDefinition, conditions, columnNames) {

	var THIS = this;

	var columns = columnNames.map(function (columnName) { return tableDefinition[columnName]; });

	return Q().then(function () {

		var queryObject = tableDefinition.select(columns.length ? columns : undefined);

		if (conditions) {

			queryObject = queryObject.where(conditions);
		}

		return THIS.queryGenerated(queryObject);
	});
};


progres.ProgresClient.prototype.selectOne = function (tableDefinition, conditions, columnNames) {

	return this
		.select(tableDefinition, conditions, columnNames)
		.then(function (rows) { return rows[0]; });
};


progres.ProgresClient.prototype.update = function (tableDefinition, conditions, object) {

	var THIS = this;

	return Q().then(function () {
	
		return THIS.queryGenerated(
			tableDefinition
				.update(object)
				.where(conditions)
		);
	});
};


progres.ProgresClient.prototype.upsert = function (tableDefinition, conditions, object) {

	return this.insert(tableDefinition, object)
		.catch(this.update(tableDefinition, conditions, object));
};


progres.ProgresClient.prototype.delete = function (tableDefinition, conditions) {

	var THIS = this;

	return Q().then(function () {

		return THIS.queryGenerated(
			tableDefinition
				.delete()
				.where(conditions)
		);
	});
};


module.exports = progres;
